### PR TITLE
Feature: Course category breadcrumbs in course header, solves #284

### PR DESCRIPTION
**Pull request by lucaboesch**
Note this also fixes some other (indiscovered) stuff in boostnavbar.php

### CODE REVIEW

**Review by abias — CHANGES_REQUESTED** [review]
Hi @lucaboesch ,

many thanks for working on this issue and for contributing another useful Boost Union feature which solves Moodle core drawbacks.

I have reviewed your PR.
To do this, I rebased it to the latest master branch (just to overcome a conflict in CHANGES.md).

On top of that, I have pushed a "Review changes" commit to your branch which coverse the following changes:

* I raised the version in version.php. You introduced new settings, but you missed to raise the version.
* I moved the breadcrumb settings from the "Look" settings page to the "Feel" settings page into the Navigation tab. I hope that you agree with me that this feature aligns well with the other existing navigation settings and does not need a tab on its own.
* I added some comments as references to the original code which you copied and modified from other locations.
* I made some (minor) code style alignments.
* I removed the `defined('MOODLE_INTERNAL') || die();` line from the boostnavbar.php class file as this is not needed in class files anymore nowadays.
* I removed the `require_once($CFG->libdir . '/navigationlib.php');` line from the boostnavbar.php class file as you have duplicated the code from there and the code worked without including the library in my tests. Please correct me if I am wrong.
* I changed your boostnavbar class in Boost Union from a class on its own to a extension of the existing boostnavbar class in Boost which means that we only have to include the functions which are changed and do not ship any unnecessary duplicated code.
* I combined your two Behat scenarios into one scenario outline. And I moved the creation of the additional categories and courses from the background into the scenario as it is not needed for all scenarios in this feature file.

-----

After that, I had a closer look at the code and tried to understand how it differs from Moodle core behaviour.
Testing on Moodle 4.1, I had to learn that Boost in Moodle core somehow behaves strangely:

* Assume that you have a category structure like this:
```
Category A -> Course 1
   Category AB -> Course 2
      Category ABC -> Course 3
          Category ABCD -> Course 4
              Category ABCDE -> Course 5
```

then:
* In Course 1, you won't see a category breadcrumb at all
* In Course 2, you won't see a category breadcrumb at all
* In Course 3, you won't see a category breadcrumb at all
* In Course 4, you _will_ see a category breakcrumb, but it says 'Category A / Category AB / Category ABC' which means that the reference to category ABCD which is the parent category of the course is missing completely. Furthermore, 'Category ABC' is just a text and not a link.
* In Course 5, you _will_ see a category breakcrumb, but it says 'Category A / Category AB / Category ABC / Category ABCD' which means that the reference to category ABCDE which is the parent category of the course is missing completely. Furthermore, 'Category ABCD' is just a text and not a link.

This feels like a Moodle core bug or at least an inconsistency. Did you mean this when you wrote that you fixed (indiscovered) stuff in boostnavbar.php as well? Do you know if this issue has been reported in Moodle tracker for Moodle core already?

-----

Anyway, as this is not about Moodle core bugs but about this feature, I tested again with Boost Union. 

When your new setting was _disabled_, the results were:

* In Course 1, you won't see a category breadcrumb at all
* In Course 2, you won't see a category breadcrumb at all
* In Course 3, you won't see a category breadcrumb at all
* In Course 4, you will see a category breakcrumb and it says 'Category A / Category AB / Category ABC' (which means that the category ABCD is still missing), but 'Category ABC' now has a link which differs from Boost core.
* In Course 5, you will see a category breakcrumb and it says 'Category A / Category AB / Category ABC / Category ABCD' (which means that the category ABCDE is still missing), but 'Category ABCD' now has a link which differs from Boost core.

After enabling your setting with the same category and course structure, the results were

* In Course 1, you still won't see a category breadcrumb at all
* In Course 2, you will see a category breakcrumb which says 'Category A / Category AB' and all items are links.
* In Course 3, you will see a category breakcrumb which says 'Category A / Category AB / Category ABC' and all items are links.
* In Course 4, you will see a category breakcrumb which says 'Category A / Category AB / Category ABC / Category ABCD' and all items are links.
* In Course 5, you will see a category breakcrumb which says 'Category A / Category AB / Category ABC / Category ABCD / Category ABCDE' and all items are links.

Looking at these test results, I am wondering if you could make three changes to your current code:

1. I would appreciate if really all courses would get a category breadcrumb if the feature is enabled. This should include Course 1 which has just one parent category
2. I would appreciate if you could extend your Behat test to cover courses on multiple levels in the category tree like in my example above. This would make the tests more secure.
3. I am wondering if we should introduce a third setting option besides "Yes" and "No" which would be called "Change nothing" and does not change anything. This is based on the fact that even if you leave the setting off, the new code behaves differently from Moodle core code. As soon as this option is introduces, the option "No" should behave in a way that the inconsistency of Moodle core is solved and the category breadcrumbs are always hidden regardless how deep the course is in the category tree.

What do you think?

Many thanks again for your contribution,
Alex

**Comment by lucaboesch** [comment]
Thanks, @abias for your insightful peer review and your interesting discoveries.
I will build something around this.
Note I'll have to rename those example course categories "Category A, Category AB, …  Category ABCDE" for the automated tests since by the nature of Behat, e.g. searching for "Category AB" unfortunately matches when "Category ABCDE" is seen on the screen, and I want to circumvent such side effects.
Best,
Luca

File: classes/boostnavbar.php
Comment (lucaboesch) [reply classes/boostnavbar.php:154]: Yes, right.
I just wanted to prevent the last item link function being removed.
Probably with the new "Change nothing" I have to leave this in and make sure it is still done when no divergent behaviour in regards to Boost core is desired.
I hesitated to make "Remove breadcrumb navigation lase item link" "yes/no" a configurable option by the way but then decided against it. This can still be done later on.

**Comment by abias** [comment]
Hi Luca,

> Note I'll have to rename those example course categories "Category A, Category AB, … Category ABCDE" for the automated tests since by the nature of Behat, e.g. searching for "Category AB" unfortunately matches when "Category ABCDE" is seen on the screen, and I want to circumvent such side effects.

Sure :)
I just meant that it would be good to test several nested categories, but the category names can be selected by you as needed.

----

I saw that you already went ahead and have added to "Do nothing" option and more test steps. Thanks for that!

Thanks for adding the new setting to the README.md file as well! I completely missed that in my review, shame on me.

But I also saw that the tests in your latest commit didn't pass.
I will have a closer look at this now, hopefully I can find the problem.

Cheers,
Alex

**Comment by abias** [comment]
Hi again,

I found a small glitch in your tests which kept them from passing.
I fixed the glitch and made some last small refinements.

As soon as the tests pass once more, the PR can be merged.

Cheers,
Alex

**Review by abias — APPROVED** [review]

**Review by abias — APPROVED** [review]

**Comment by abias** [comment]
This is merged now, Luca!

You can now advertise the solution in https://moodle.org/mod/forum/discuss.php?d=433872 and https://tracker.moodle.org/browse/MDL-77934 if you like :)

**Comment by lucaboesch** [comment]
On Moodle Tracker, [MDL-78138](https://tracker.moodle.org/browse/MDL-78138) is now in "Waiting for Integration Review" status.

**Pull request merged**
